rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,572 @@
+
+To build your own modal dev tool, simply create a configuration resource, implement the `ModalDevTool` trait, and then register it in the app when the correct feature flag is enabled.
+
+```rust
+/// A flying camera controller that lets you disconnect your camera from the player to freely explore the environment.
+/// 
+/// When this mode is disabled
+#[derive(Resource, Reflect, FromReflect, Debug)]
+struct DevFlyCamera {
+    enabled: bool,
+    /// How fast the camera travels forwards, backwards, left, right, up and down, in world units.
+    movement_speed: Option<f32>,
+    /// How fast the camera turns, in radians per second.
+    turn_speed: Option<f32>,
+}
+
+impl Default for DevFlyCamera {
+    fn default() -> Self {
+        DevFlyCamera {
+            enabled: false,
+            movement_speed: 3.,
+            turn_speed: 10.
+        }
+    }
+}
+
+impl ModalDevTool for DevFlyCamera {
+    /// Expects a call that looks like `dev_fly_camera` (default settings)
+    /// or `dev_fly_camera 5 10` (to specify settings)
+    fn parse_from_str(string: &str) -> Result<Self, DevToolParseError>{
+        let parts = string.split_whitespace();
+        let name = parts.iter.next()?;
+        if name != self.name() {
+            return Err(DevToolParseError::InvalidName);
+        }
+
+        let movement_speed: f32 = match parts.iter.next() {
+            Some(string) = parse(string)?,
+            None => Self::default().movement_speed,
+        };
+
+        let turn_speed: f32 = match parts.iter.next() {
+            Some(string) = parse(string)?,
+            None => Self::default().turn_speed,
+        };
+
+        Ok(DevFlyCamera {
+                enabled: true,
+                movmeent_speed,
+                turn_speed,
+            }
+        )
+    }
+
+    fn set_enabled(&mut self, enabled: bool) {
+        self.enabled = enabled;
+    }
+
+    fn is_enabled(&self) -> bool {
+        self.enabled
+    }
+}
+```
+
+To configure a dev tool at runtime, simply access the resource, and either mutate or overwrite the `ModalDevTools` struct.
+
+### Dev commands
+
+Dev commands modify the world a single time, and can be called with arguments.
+To model this, we leverage Bevy's existing `Command` trait, which exists to perform complex one-off modifications to the world.
+
+```rust
+/// Dev commands are used by developers to modify the `World` in order to easily debug and test their application.
+/// 
+/// Dev commands can be called with arguments to specify the exact behavior: if you are creating a toolbox, parse the provided arguments 
+/// to construct an instance of the type that implements this type, and then send it as a `Command` to execute it.
+/// 
+/// The documentation on this struct is reflected, and can be read by toolboxes to provide help text to users.
+trait DevCommand: Command + Reflect + FromReflect + Debug + 'static {
```

Comment:
Instead of the `DevCommand` trait including a `parse_from_str` method, it would make more sense to require `DevCommand: FromStr`.